### PR TITLE
Simplified URLs configuration

### LINKED ISSUE #13: Simplify URLs configuration
Let's think how to simplify the configuration of URL regexps. Right now, you need to write such things:

```js
url: [
	/^((https?:)?\/\/)?(www.)?dailymotion\.com\/video\/(\w+)/,
],
```

That's awfully long and complicated. Could we remove the protocol and the usually optional `www.` piece from the config? But at the same time, how to do that in a way which won't break if someone configured this feature using a regexp containing the whole URL?

When the user provides us a URL we need to use it for the below jobs:

* To match it with the config (where we're matching with either a full or partial URL regexp).
* [FUTURE] When the user pastes into the content ([autoembed](https://github.com/ckeditor/ckeditor5-media-embed/issues/6)) where it needs to be picked up as a whole (including the protocol) even if the feature is configured using partial URL regexps.
* To print it in `data-oembed-url` (where a full URL should be output).

## Algorithms

For matching with the config:

 1. Try matching without stripping the protocol.
 1. If doesn't work, try stripping the protocol from the passed URL (`https://www.youtube.com/foo` -> `youtube.com/foo`) and match with the config again.

For producing `data-oembed-url`:

1. Check if starts with protocol
1. If not, add `https://`
1. If anyone will report that they don't want `https://`, then the workaround is – make sure users pass the entire URL and we can make this bit of behaviour configurable too. But KISS for now – let's always ensure that `https://` is there.

For pasting URLs (autoembed):

1. Use a general URL regexp to find out whether someone pasted an URL (and match its entire length).
1. Match that with the config as in the first scenario.

## Config

Finally, the config will look like this:

```js
url: [
  /^dailymotion\.com\/video\/(\w+)/,                            // RECOMMENDED
  /^((https?:)?\/\/)?(www.)?dailymotion\.com\/video\/(\w+)/,    // NOT RECOMMENDED, but works too.
],
```

### LINKED ISSUE #26: Wrong Regex for youtube video
For youtube id you have \w+ pattern which is equal to [a-zA-Z0-9_]+. But youtube links can have "-" character in id.
For example if I paste link https://www.youtube.com/watch?v=euqbMkM-QQk data-oembed-url will be correct while iframe src attribute will be https://www.youtube.com/embed/euqbMkM

### PR DESCRIPTION
**Pull request by pomek**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Feature: Simplified URLs configuration. Now RegExps do not have defined a protocol. Changed argument passed to `MediaEmbedProvider#html` function. Instead of the last regexp match, it will receive the entire matched array. Changed a RegExp for video from YouTube because media that contained a dash in URL weren't parsed correctly. Closes #13. Closes #26.

### CODE REVIEW

File: src/mediaembedediting.js [hunk 1]
```diff
@@ -34,88 +34,108 @@ export default class MediaEmbedEditing extends Plugin {
...
-					html: id =>
-						'<div style="position: relative; padding-bottom: 100%; height: 0; ">' +
-							`<iframe src="https://www.dailymotion.com/embed/video/${ id }" ` +
-								'style="position: absolute; width: 100%; height: 100%; top: 0; left: 0;" ' +
-								'frameborder="0" width="480" height="270" allowfullscreen allow="autoplay">' +
-							'</iframe>' +
-						'</div>'
+					url: /^dailymotion\.com\/video\/(\w+)/,
+					html: match => {
+						const id = match.slice( -1 );
```
Comment (Reinmar) [inline src/mediaembedediting.js]: You know exactly at which index of `match` this capturing group is available, so why complicating this with `.slice( -1 )`?

**Comment by Reinmar** [comment]
Where are the tests for the core of this change? I see 0 new tests while you added quite a lot of new functionality.

**Comment by Reinmar** [comment]
> IMO they aren't needed because I must updated existing tests to pass and they already checked everything.

100% CC doesn't mean that the code is covered. For instance, when you have a regexp in your code a single test can cover that line, but does it cover entire regexp? Is a single test able to validate multiple patterns which that regexp must match?

Secondly, tests should document the code. They should explain what the code does. But you wrote 30LOC and didn't add any test. The previously existing tests could not document this code because they existed before you wrote it so they clearly tested something else. Which means that this feature is not documented by the tests.

You can also think about this situation in this way – I shouldn't be able to break your implementation by rewriting it in a lousy way (or by implementing a different algorithm) without making the tests fail. Do you wanna bet? ;>

**Comment by pomek** [comment]
Existing tests cover my changes. Do I have to copy it and add once again to "cover my changes"?

File: src/mediaembedediting.js [hunk 2]
```diff
@@ -34,88 +34,108 @@ export default class MediaEmbedEditing extends Plugin {
...
 				},
 
 				{
 					name: 'youtube',
 					url: [
-						/^(https:\/\/)?(www\.)?youtube\.com\/watch\?v=(\w+)/,
-						/^(https:\/\/)?(www\.)?youtube\.com\/v\/(\w+)/,
-						/^(https:\/\/)?(www\.)?youtube\.com\/embed\/(\w+)/,
-						/^(https:\/\/)?youtu\.be\/(\w+)/
+						/^youtube\.com\/watch\?v=([a-zA-Z0-9_-]+)/,
```
Comment (Reinmar) [inline src/mediaembedediting.js]: Why not `[\w-]`?

Comment (pomek) [reply src/mediaembedediting.js]: Simplified. Thx.

**Comment by Reinmar** [comment]
Existing tests cover your changes by coincident. Only because they are high-level enough to trigger the code that you wrote and contain enough variations to cover all `if()s`.

Let's try to look at this from a different angle – would someone else be able to implement an algorithm compatible with what you implemented if we removed the entire code and left only the tests? Because we can also say that tests are a specification for the implementation. And you haven't changed a single test description which means that you haven't written that specification.

**Comment by pomek** [comment]
> And you haven't changed a single test description which means that you haven't written that specification.

Again. Because these tests are high-level, they are correct before and after changes.

But I understand your point of view before we started the discussion. Because of the lack of tests for `MediaRegistry` class, I don't have to change too many things.

**Comment by Reinmar** [comment]
> Should I create tests for this class?

That's one of the solutions :) You're right that what's needed now are lower-level tests – they can be e.g. `MediaRegistry`'s tests – that's a good idea.

**Comment by pomek** [comment]
I've just added some basic tests for the feature that we would like to add in this PR.

File: src/mediaregistry.js [hunk 3]
```diff
@@ -284,4 +319,22 @@ class Media {
...
+	 *
+	 * @param {String} url The url of the media.
+	 * @returns {String|null}
+	 */
+	_getValidUrl( url ) {
+		if ( !url ) {
+			return null;
+		}
+
+		if ( url.match( /^https?/ ) ) {
```
[the remark this answers is not in the document]
Comment (pomek) [reply src/mediaregistry.js]: `if ( url.match( /^https?/ ) ) {` – accepts `http` and `https` as well.

[the remark this answers is not in the document]
Comment (pomek) [reply src/mediaregistry.js]: > I noticed that for protocol-less URLs the preview renders the "open in new tab" starting with "https://"?

It's by design. Described in the ticket.

File: tests/mediaregistry.js [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+			expect( availableProviders ).to.deep.equal( [ 'spotify', 'youtube', 'vimeo' ] );
+			expect( logStub.calledOnce ).to.equal( true );
+			expect( logStub.firstCall.args[ 0 ] ).to.equal(
+				'media-embed-no-provider-name: The configured media provider has no name and cannot be used.'
+			);
+			expect( logStub.firstCall.args[ 1 ] ).to.deep.equal( { provider: { url: [ /dailymotion\.com/ ] } } );
+		} );
+	} );
+
+	describe( '_getMedia()', () => {
```
[the remark this answers is not in the document]
Comment (pomek) [reply tests/mediaregistry.js]: You're saying one thing, @Reinmar said about missing tests for the changes… Sorry, I don't get it :(

[the remark this answers is not in the document]
Comment (pomek) [reply tests/mediaregistry.js]: Is changing it to protected will be enough?

**Comment by pomek** [comment]
I did some fixes according to your comments.

One thing is still missing:

> Also: what do we do with protocol-relative URLs (starting with //)? Are we considering such media?

File: tests/mediaregistry.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (oleq) [reply tests/mediaregistry.js]: I'm not saying testing is wrong ;-) Just let's keep it consistent with our policies. Yeah, `@protected` is enough.

File: src/mediaregistry.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (oleq) [reply src/mediaregistry.js]: Moved the discussion about protocol–relative URLs to https://github.com/ckeditor/ckeditor5-media-embed/issues/33.

**Comment by oleq** [comment]
@pomek Moved it to the follow–up https://github.com/ckeditor/ckeditor5-media-embed/issues/33. Let's not waste our time on that because probably it's an edge–case.

**Comment by pomek** [comment]
Ok. So I guess you can review it once again.

**Pull request merged**